ws/ShardManager.ts

```diff
@@ -0,0 +1,34 @@
+import DiscordWebsocket, { ConnectionProperties } from './DiscordWebsocket';
+
+interface BotGateway {
+  session_start_limit: {
+    max_concurrency: number;
+    remaining: number;
+    reset_after: number;
+    total: number;
+  };
+  shards: number;
+  url: string;
+}
+
+export default class ShardManager {
+  shards: Map<number, DiscordWebsocket> = new Map();
+
+  constructor(public token: string, public intents: number, public connProps: Partial<ConnectionProperties> = {}, public botGatewayURL: string = 'https://discord.com/api/v8/gateway/bot') {
+    this.initialize();
+  }
+
+  private async getBotGateway(): Promise<BotGateway> {
```

Comment:
I'm wondering if we can use our REST client here

Reply:
we'll see if we can abstract some part of the request handler in its own subpackage.
A basic req handler that handles ratelimits for anything discord could be a nice thing. The rest package would be the impl with like all implemented methods and endpoints.
However that's not something we'll want nor need in this PR